eureka-core/src/main/java/com/netflix/eureka/registry/AbstractInstanceRegistry.java

```diff
@@ -1236,67 +1215,11 @@ public boolean offer(E e) {
-        // The OUT_OF_SERVICE from the client or replica needs to be confirmed
-        // as well since the service may be
-        // currently in SERVICE
-        if (
-                (!InstanceStatus.UP.equals(r.getStatus()))
-                        && (!InstanceStatus.OUT_OF_SERVICE.equals(r.getStatus()))) {
-            logger.debug("Trusting the instance status {} from replica or instance for instance {}",
-                    r.getStatus(), r.getId());
-            return r.getStatus();
-        }
-        // Overrides are the status like OUT_OF_SERVICE and UP set by NAC
-        InstanceStatus overridden = overriddenInstanceStatusMap.get(r.getId());
-        // If there are instance specific overrides, then they win - otherwise the ASG status
-        if (overridden != null) {
-            logger.debug("The instance specific override for instance {} and the value is {}",
-                    r.getId(), overridden.name());
-            return overridden;
-        }
-        // If the ASGName is present- check for its status
-        boolean isASGDisabled = false;
-        if (r.getASGName() != null) {
-            isASGDisabled = !AwsAsgUtil.getInstance().isASGEnabled(r);
-            logger.debug("The ASG name is specified {} and the value is {}", r.getASGName(), isASGDisabled);
-            if (isASGDisabled) {
-                return InstanceStatus.OUT_OF_SERVICE;
-            } else {
-                return InstanceStatus.UP;
-            }
-        }
-        // This is for backward compatibility until all applications have ASG
-        // names, otherwise while starting up
-        // the client status may override status replicated from other servers
-        if (!isReplication) {
-            InstanceStatus existingStatus = null;
-            if (existingLease != null) {
-                existingStatus = existingLease.getHolder().getStatus();
-            }
-            // Allow server to have its way when the status is UP or OUT_OF_SERVICE
-            if (
-                    (existingStatus != null)
-                            && (InstanceStatus.OUT_OF_SERVICE.equals(existingStatus)
-                            || InstanceStatus.UP.equals(existingStatus))) {
-                logger.debug("There is already an existing lease with status {}  for instance {}",
-                        existingLease.getHolder().getStatus().name(),
-                        existingLease.getHolder().getId());
-                return existingLease.getHolder().getStatus();
-            }
-        }
-        logger.debug("Returning the default instance status {} for instance {}",
-                r.getStatus(), r.getId());
-        return r.getStatus();
-    }
-
+    // FIXME this should really be part of peerAware registry as it has knowledge of replication ...
+    // fix in next round of code clean up
+    protected abstract InstanceInfo.InstanceStatus getOverriddenInstanceStatus(InstanceInfo r,
```

Comment:
The replication flag comes from REST API call, and we pass it down from there. I think it is fine to have this logic here.